### PR TITLE
Multi Volume optimization and boundary check

### PR DESCRIPTION
**Pull request by Luvideria**
This PR updates Multi volumes with boundary check and removes one loop as well as local vector storage.

With hundred spheres, the speedup is approximately 10%
Further improvement is expected by using a volume datastructure.

### CODE REVIEW

File: src/volume/volume_multi.cpp [hunk 1]
```diff
@@ -90,11 +90,25 @@ class Volume_Multi : public Volume {
         return max_scalar_;
     }
 
+    bool isInBound(const Vec3& p, const Bound& b) const {
```
Comment (hi2p-perim) [inline src/volume/volume_multi.cpp:93]: Naming: please use snake case `is_in_bound` for the function name.

File: src/volume/volume_multi.cpp [hunk 2]
```diff
@@ -90,11 +90,25 @@ class Volume_Multi : public Volume {
...
+        auto isBetween=[](float a, float low, float high)->bool{
```
Comment (hi2p-perim) [inline src/volume/volume_multi.cpp:94]: `const auto is_between = [](float a, float low, float high) -> bool {`

File: src/volume/volume_multi.cpp [hunk 3]
```diff
@@ -90,11 +90,25 @@ class Volume_Multi : public Volume {
...
+        return isBetween(p.x,b.min.x,b.max.x) &&
+               isBetween(p.y,b.min.y,b.max.y) &&
+               isBetween(p.z,b.min.z,b.max.z);
+    }
+    
     // Computes the sum over all Volumes of eval_scalar
     virtual Float eval_scalar(Vec3 p) const override {
         Float sum = 0._f;
         for(auto* v : volumes_den_)
+        {
```
Comment (hi2p-perim) [inline src/volume/volume_multi.cpp:107]: Please move `{` to the end of previous line.

File: src/volume/volume_multi.cpp [hunk 4]
```diff
@@ -103,26 +117,21 @@ class Volume_Multi : public Volume {
...
         Float sum = 0;
-        scalars.reserve(size_);
-
-        // Perform a prefix sum of scalar evaluations
-        for(auto* v : volumes_den_){
-            Float sc = v->eval_scalar(p);
-            scalars.push_back(sc);
+        Vec3 resulting_color(0._f);
+        
+        for(unsigned int i = 0; i < size_; i++){
```
Comment (hi2p-perim) [inline src/volume/volume_multi.cpp:125]: Please add a space between `for` and parenthesis, and parenthesis and `{`.

File: src/volume/volume_multi.cpp [hunk 5]
```diff
@@ -103,26 +117,21 @@ class Volume_Multi : public Volume {
...
+            if( !isInBound(p,volumes_den_[i]->bound()) )
```
Comment (hi2p-perim) [inline src/volume/volume_multi.cpp:126]: `if (!is_in_bound(p, volumes_den_[i]->bound()))`

**Comment by Luvideria** [comment]
Maybe we could make a cosmetic lint guide to put in the doc? I keep messing up. I think there are some auto-prettifyer that one can configure, I don't know if it's worth the effort though. It could probably handle simple cases such as spaces after comma, space after if, spaces inside for parenthesis and )[space]{\n

**Comment by hi2p-perim** [comment]
That makes sense. I will make a doc about the coding convention about the project. Generally, I like to follow [Google C++ Style guide](https://google.github.io/styleguide/cppguide.html) but there's small differences since the project is a hybrid Python/C++ library.

**Pull request merged**